### ISSUE TITLE
two ocaml <-> lib torch syntax questions

### ISSUE THREAD

**Issue by zeroexcuses** [LaurentMazare/ocaml-torch#16]
Here is a small example for doing back prop:

```
module Tch = Torch;;
module Ten = Torch.Tensor ;;
module Tchc = Torch_core ;;

let x = Ten.ones [2; 2] ~requires_grad:true ~kind:Tchc.Kind.Float in
let y = Ten.(x + Ten.f 2.) in
let z = Ten.(y * y * f 3.) in
let out = Ten.mean z in

Ten.grad x ;;

```

There are two things I find slightly annoying about this code and want to see if there is a way to fix:

1. the `let x = ... in` expressions. I don't mind typing in `let x = ...`, as we do that all the time in Rust. However, in Rust, we do `let x = ... ;` it's a single line on it's own, defining a var in the local context. In OCaml, it seems that every `let ... in` is creating a new environment, containing precisely one binding. Is there a way around this, or is this a fundamental OCaml design ?

2. The other thing I don't quite like is this line here: `let _ = Ten.backward out  in` when I first look at this, I'm expecting `Ten.backward` to be a pure function and to be returning a result. But what' it's doing is modifying the memory associated with the gradients for `out` (and nodes it depends on). If it was something like `out.do_backward_pass()` it would be a bit more obvious that it's modifying the Tensor. Is there a way around this, or is this also the way it is when using OCaml ?

In particular, I'm expecting (perhaps this is unreasonable -- and I'm overly sensitive to this due to Rust's ref vs mut ref semantics) for functions to be pure, and for things that modify to be member functions.

**Comment by zeroexcuses** [comment]
Understood, thanks!
